lib/tree.rb

```diff
@@ -1,25 +1,48 @@
 class NoApplesError < StandardError; end
 
 class Tree
-  attr_#fill_in :height, :age, :apples, :alive
+  attr_accessor :height, :age, :apples, :alive
 
   def initialize
+    self.height ||= 0
+    self.age ||= 0
+    self.apples ||= []
+    self.alive = true
   end
 
   def age!
+    if self.age < 50
+      self.height += 1
+    end
+
+    if self.age == 99
+      self.alive = false
+    else
+      add_apples
+    end
+
+    self.age += 1
   end
 
   def add_apples
+    5.times do
+      diameter = Random.rand(2..4)
+      color = 'red'
+      apples[apples.length] = Apple.new(color, diameter)
+    end
   end
 
   def any_apples?
+    !apples.empty?
   end
 
   def pick_an_apple!
     raise NoApplesError, "This tree has no apples" unless self.any_apples?
+    apples.shift
   end
 
   def dead?
+    !self.alive
```

Comment:
You are consistently using the `self.attribute` for setting/changing values and just the bare `attribute` for reading values, except this one reader. Maybe change this one to `!alive`?

Reply:
@mikegee what's your preference for defining attributes? ivars or using self as done here? And why? :D

Reply:
I prefer defining the attribute methods and using them exclusively. Instance variables with typos raise no error, Ruby happily returns `nil`. Method calls with typos raise `NoMethodError`. I prefer feedback as early as possible.

Reply:
On this particular line, I suggest removing the `self.` prefix but continuing to call the reader method.